modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -65,3 +78,85 @@ object GlobalSnapshotStorage {
   case class InvalidGlobalSnapshotChain(expected: GlobalSnapshotLink, actual: GlobalSnapshotLink) extends NoStackTrace
 
 }
+
+object GlobalSnapshotStorageV2 {
+
+  def make[F[_]: Async](
+    globalSnapshotLocalFileStorage: GlobalSnapshotLocalFileStorage[F],
+    genesis: GlobalSnapshot,
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorageV2[F]] = {
+    def mkCacheQueue = Queue.unbounded[F, GlobalSnapshot]
+    def mkHeadRef = Ref.of[F, GlobalSnapshot](genesis)
+    def mkCache = MapRef.ofSingleImmutableMap[F, SnapshotOrdinal, GlobalSnapshot](Map.empty)
+    def mkOffloadQueue = Queue.unbounded[F, SnapshotOrdinal]
+
+    def mkLogger = Slf4jLogger.create[F]
+
+    mkLogger.flatMap { implicit logger =>
+      (mkCacheQueue, mkHeadRef, mkCache, mkOffloadQueue).mapN {
+        make(_, _, _, _, globalSnapshotLocalFileStorage, inMemoryCapacity)
+      }.flatten
+    }
+  }
+
+  def make[F[_]: Async: Logger](
+    cacheQueue: Queue[F, GlobalSnapshot],
+    headRef: Ref[F, GlobalSnapshot],
+    cache: MapRef[F, SnapshotOrdinal, Option[GlobalSnapshot]],
+    offloadQueue: Queue[F, SnapshotOrdinal],
+    globalSnapshotLocalFileStorage: GlobalSnapshotLocalFileStorage[F],
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorageV2[F]] = {
+
+    def cacheProcess =
+      Stream
+        .fromQueueUnterminated(cacheQueue)
+        .evalTap { snapshot =>
+          headRef.set(snapshot) >>
+            cache(snapshot.ordinal).set(snapshot.some) >>
+            NonNegLong
+              .from(snapshot.ordinal.value.value - inMemoryCapacity.value)
+              .map(SnapshotOrdinal(_))
+              .fold(_ => Applicative[F].unit, offloadQueue.offer)
+        }
+        .compile
+        .drain
+
+    def offloadProcess =
+      Stream
+        .fromQueueUnterminated(offloadQueue)
+        .evalTap { ordinal =>
+          cache(ordinal).get.flatMap {
+            case Some(snapshot) =>
+              globalSnapshotLocalFileStorage.write(snapshot).rethrowT.handleErrorWith { e =>
+                Logger[F].error(e)(s"Failed writing global snapshot to disk! Snapshot ordinal=${snapshot.ordinal.show}")
+              } >> cache(ordinal).set(none)
+            case None => MonadThrow[F].raiseError[Unit](new Throwable("Unexpected state!"))
+          }
+        }
+        .compile
+        .drain
+
+    Spawn[F].start { offloadProcess } >> Spawn[F].start { cacheProcess }.map { _ =>
+      new GlobalSnapshotStorageV2[F] {
+        def prepend(snapshot: GlobalSnapshot) =
+          head.map { _.ordinal.next === snapshot.ordinal }.flatTap {
```

Comment:
I think we can use `ifM`

Reply:
Why? I don't want to flatMap but flatTap

Reply:
Oh, ok you are returning it to the caller.